File: server/Jellyfin.Plugin.Tally/Dvr/DvrModels.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Serialization;
using Jellyfin.Plugin.Tally.Scores;

namespace Jellyfin.Plugin.Tally.Dvr;

/// <summary>Where a recording job is. Scheduled (game upcoming) → waiting (for a stream, or a free slot) → recording →
/// finishing (remux) → done, failed (with a plain reason) or canceled.</summary>
public static class JobState
{
    public const string Scheduled = "scheduled";
    public const string Waiting = "waiting";
    public const string Recording = "recording";
    public const string Finishing = "finishing";
    public const string Done = "done";
    public const string Failed = "failed";
    public const string Canceled = "canceled";

    public static bool IsFinal(string state) => state is Done or Failed or Canceled;

    /// <summary>Not started yet: canceling it records nothing.</summary>
    public static bool IsPending(string state) => state is Scheduled or Waiting;
}

/// <summary>DVR settings, server-wide. Kept in the DVR's own file (not the plugin configuration), so the settings
/// page's configuration save can never write back stale DVR values.</summary>
public sealed class DvrSettings
{
    public const long Gb = 1024L * 1024 * 1024;

    /// <summary>Recordings folder; empty = a "Tally Recordings" folder next to Jellyfin's data folder.</summary>
    [JsonPropertyName("folder")] public string Folder { get; set; } = string.Empty;

    /// <summary>Free space to leave on the recordings drive: a recording does not start, and stops, below it.</summary>
    [JsonPropertyName("reserveBytes")] public long ReserveBytes { get; set; } = 10 * Gb;

    [JsonPropertyName("maxConcurrent")] public int MaxConcurrent { get; set; } = 3;

    /// <summary>Minutes recorded after the scoreboard says the game is final.</summary>
    [JsonPropertyName("postRollMinutes")] public int PostRollMinutes { get; set; } = 5;

    [JsonPropertyName("maxHours")] public double MaxHours { get; set; } = 6;

    /// <summary>Delete recordings this many days after they were recorded; 0 = never.</summary>
    [JsonPropertyName("deleteAfterDays")] public int DeleteAfterDays { get; set; }

    public DvrSettings Normalized()
    {
        return new DvrSettings
        {
            Folder = (Folder ?? string.Empty).Trim(),
            ReserveBytes = Math.Clamp(ReserveBytes, 0, 100_000 * Gb),
            MaxConcurrent = Math.Clamp(MaxConcurrent, 1, 20),
            PostRollMinutes = Math.Clamp(PostRollMinutes, 0, 120),
            MaxHours = Math.Clamp(MaxHours, 0.25, 24),
            DeleteAfterDays = Math.Clamp(DeleteAfterDays, 0, 3650)
        };
    }
}

/// <summary>One side of a recorded game, as the recording remembers it. Deliberately without a score: nothing the
/// DVR writes (job file, file name, NFO, art) can carry the result.</summary>
public sealed class TeamSnapshot
{
    [JsonPropertyName("id")] public string Id { get; set; } = string.Empty;

    [JsonPropertyName("abbr")] public string Abbr { get; set; } = string.Empty;

    [JsonPropertyName("name")] public string Name { get; set; } = string.Empty;

    [JsonPropertyName("shortName")] public string ShortName { get; set; } = string.Empty;

    [JsonPropertyName("nickname")] public string Nickname { get; set; } = string.Empty;

    [JsonPropertyName("location")] public string Location { get; set; } = string.Empty;

    [JsonPropertyName("logo")] public string Logo { get; set; } = string.Empty;

    [JsonPropertyName("color")] public string Color { get; set; } = string.Empty;

    [JsonPropertyName("altColor")] public string AltColor { get; set; } = string.Empty;

    public static TeamSnapshot From(GameTeam t) => new()
    {
        Id = t.Id, Abbr = t.Abbr, Name = t.Name, ShortName = t.ShortName, Nickname = t.Nickname,
        Location = t.Location, Logo = t.Logo, Color = t.Color, AltColor = t.AltColor
    };

    public GameTeam ToTeam() => new()
    {
        Id = Id, Abbr = Abbr, Name = Name, ShortName = ShortName, Nickname = Nickname,
        Location = Location, Logo = Logo, Color = Color, AltColor = AltColor
    };

    /// <summary>The name a title uses: the full name, else the short one, else the abbreviation.</summary>
    [JsonIgnore]
    public string DisplayName => !string.IsNullOrWhiteSpace(Name) ? Name : !string.IsNullOrWhiteSpace(ShortName) ? ShortName : Abbr;
}

/// <summary>The game a job records, frozen when the job was made and refreshed from the scoreboard while it runs
/// (start time, teams), never with a score.</summary>
public sealed class GameSnapshot
{
    [JsonPropertyName("id")] public string Id { get; set; } = string.Empty;

    [JsonPropertyName("leaguePath")] public string LeaguePath { get; set; } = string.Empty;

    [JsonPropertyName("league")] public string League { get; set; } = string.Empty;

    [JsonPropertyName("sport")] public string Sport { get; set; } = string.Empty;

    [JsonPropertyName("start")] public DateTimeOffset Start { get; set; }

    [JsonPropertyName("away")] public TeamSnapshot Away { get; set; } = new();

    [JsonPropertyName("home")] public TeamSnapshot Home { get; set; } = new();

    [JsonPropertyName("broadcasts")] public List<string> Broadcasts { get; set; } = new();

    /// <summary>"Away at Home".</summary>
    [JsonPropertyName("title")] public string Title => $"{Away.DisplayName} at {Home.DisplayName}";

    public static GameSnapshot From(GameInfo g) => new()
    {
        Id = g.Id,
        LeaguePath = g.LeaguePath,
        League = g.League,
        Sport = g.Sport,
        Start = g.Start,
        Away = TeamSnapshot.From(g.Away),
        Home = TeamSnapshot.From(g.Home),
        Broadcasts = g.Broadcasts.ToList()
    };

    /// <summary>A scoreless, not-yet-started game for the art renderers.</summary>
    public GameInfo ToArtGame() => new()
    {
        Id = Id, League = League, Sport = Sport, Start = Start, State = "pre", LeaguePath = LeaguePath,
        Away = Away.ToTeam(), Home = Home.ToTeam(), Broadcasts = Broadcasts.ToList(), Name = Title
    };
}

/// <summary>What to record: one game, or every game of a team (optionally in one league). Server-wide.</summary>
public sealed class RecordingRule
{
    public const string GameKind = "game";
    public const string TeamKind = "team";

    [JsonPropertyName("id")] public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("kind")] public string Kind { get; set; } = GameKind;

    /// <summary>Game rules: the game's league; team rules: the league it is limited to, empty = any league.</summary>
    [JsonPropertyName("leaguePath")] public string LeaguePath { get; set; } = string.Empty;

    [JsonPropertyName("gameId")] public string? GameId { get; set; }

    [JsonPropertyName("teamId")] public string? TeamId { get; set; }

    [JsonPropertyName("teamName")] public string? TeamName { get; set; }

    [JsonPropertyName("title")] public string Title { get; set; } = string.Empty;

    /// <summary>Team rules: keep the last N recorded games, 0 = all.</summary>
    [JsonPropertyName("keepLast")] public int KeepLast { get; set; }

    [JsonPropertyName("createdBy")] public Guid CreatedBy { get; set; }

    [JsonPropertyName("createdByName")] public string CreatedByName { get; set; } = string.Empty;

    [JsonPropertyName("createdAt")] public DateTimeOffset CreatedAt { get; set; }

    /// <summary>Team rules: this game is one of the team's (in the rule's league, if it has one).</summary>
    public bool Covers(GameInfo g)
        => Kind == TeamKind
           && !string.IsNullOrEmpty(TeamId)
           && (string.IsNullOrEmpty(LeaguePath) || string.Equals(LeaguePath, g.LeaguePath, StringComparison.OrdinalIgnoreCase))
           && (g.Home.Id == TeamId || g.Away.Id == TeamId);
}

/// <summary>One game being (or to be, or that was) recorded.</summary>
public sealed class RecordingJob
{
    [JsonPropertyName("id")] public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("ruleId")] public Guid RuleId { get; set; }

    [JsonPropertyName("game")] public GameSnapshot Game { get; set; } = new();

    [JsonPropertyName("state")] public string State { get; set; } = JobState.Scheduled;

    /// <summary>Why it is waiting, why it failed, or why it stopped early.</summary>
    [JsonPropertyName("reason")] public string? Reason { get; set; }

    /// <summary>Set when a recording stopped before the game's end: "canceled", "disk", "maxLength", "postponed".</summary>
    [JsonPropertyName("stopReason")] public string? StopReason { get; set; }

    [JsonPropertyName("createdAt")] public DateTimeOffset CreatedAt { get; set; }

    [JsonPropertyName("startedAt")] public DateTimeOffset? StartedAt { get; set; }

    [JsonPropertyName("endedAt")] public DateTimeOffset? EndedAt { get; set; }

    /// <summary>When the scoreboard first said the game was final (the post-roll counts from here).</summary>
    [JsonPropertyName("finalSeenAt")] public DateTimeOffset? FinalSeenAt { get; set; }

    [JsonPropertyName("channelId")] public string? ChannelId { get; set; }

    [JsonPropertyName("channelName")] public string? ChannelName { get; set; }

    [JsonPropertyName("bytes")] public long Bytes { get; set; }

    [JsonPropertyName("seconds")] public double Seconds { get; set; }

    [JsonPropertyName("segments")] public int Segments { get; set; }

    /// <summary>Measured bits per second of what was recorded.</summary>
    [JsonPropertyName("bitrate")] public double? Bitrate { get; set; }

    [JsonPropertyName("lastSegmentAt")] public DateTimeOffset? LastSegmentAt { get; set; }

    /// <summary>How many times the recording was picked up again after a server restart.</summary>
    [JsonPropertyName("resumes")] public int Resumes { get; set; }

    /// <summary>The finished file.</summary>
    [JsonPropertyName("filePath")] public string? FilePath { get; set; }

    [JsonPropertyName("fileBytes")] public long FileBytes { get; set; }

    /// <summary>The Jellyfin library item of the finished file, once Jellyfin has scanned it.</summary>
    [JsonPropertyName("itemId")] public string? ItemId { get; set; }

    [JsonPropertyName("workDir")] public string? WorkDir { get; set; }
}

/// <summary>Everything the DVR keeps on disk.</summary>
public sealed class DvrState
{
    [JsonPropertyName("settings")] public DvrSettings Settings { get; set; } = new();

    [JsonPropertyName("rules")] public List<RecordingRule> Rules { get; set; } = new();

    [JsonPropertyName("jobs")] public List<RecordingJob> Jobs { get; set; } = new();

    /// <summary>When the DVR created the recordings library by itself (the first finished recording without one). Set
    /// once and never cleared: a library the owner removes afterwards is not created again.</summary>
    [JsonPropertyName("libraryAutoCreatedAt")] public DateTimeOffset? LibraryAutoCreatedAt { get; set; }
}

/// <summary>A game's recording as the board shows it.</summary>
public sealed class GameRecording
{
    [JsonPropertyName("state")] public string State { get; set; } = string.Empty;

    [JsonPropertyName("jobId")] public string JobId { get; set; } = string.Empty;

    /// <summary>Root-relative, signed HLS of everything recorded so far (EVENT playlist), while recording.</summary>
    [JsonPropertyName("startOverPath")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? StartOverPath { get; set; }

    /// <summary>The library item of the finished recording.</summary>
    [JsonPropertyName("itemId")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ItemId { get; set; }

    /// <summary>"ready", "adding" or "noLibrary" (see <see cref="RecordingLibrary"/>).</summary>
    [JsonPropertyName("libraryState")] public string LibraryState { get; set; } = RecordingLibrary.NoLibrary;

    [JsonPropertyName("reason")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Reason { get; set; }
}

File: app/src/main/java/io/github/scdouglas1999/tally/dvr/DvrModels.kt
package io.github.scdouglas1999.tally.dvr

import io.github.scdouglas1999.tally.api.TallyGame
import io.github.scdouglas1999.tally.api.TallyTeam
import kotlinx.serialization.Serializable

/*
 * The server DVR's client API (`/JellyTV/Client/v1/recordings…`, plugin feature `dvr`), decoded with TallyJson
 * (lenient: unknown keys ignored, missing keys take the defaults). Nothing here ever carries a score.
 */

/** `GET /JellyTV/Client/v1/recordings`: every rule and job on the server, and whether this user may change them. */
@Serializable
data class DvrList(
    /** The Jellyfin "manage recordings" permission (admins always have it). Without it every action is absent. */
    val canManage: Boolean = false,
    val reason: String? = null,
    val rules: List<DvrRule> = emptyList(),
    /** Recording, finishing, waiting, scheduled first; then the finished ones, newest first. */
    val jobs: List<DvrJob> = emptyList(),
)

@Serializable
data class DvrRule(
    val id: String = "",
    /** game | team */
    val kind: String = "",
    val title: String = "",
    /** Game rules: the game's league; team rules: the league it is limited to, empty = any league. */
    val leaguePath: String = "",
    val gameId: String? = null,
    val teamId: String? = null,
    val teamName: String? = null,
    /** Team rules: keep the last N recorded games, 0 = all. */
    val keepLast: Int = 0,
    val createdByName: String = "",
) {
    val isTeam: Boolean get() = kind == KIND_TEAM

    companion object {
        const val KIND_TEAM = "team"
    }
}

@Serializable
data class DvrTeam(
    val id: String = "",
    val abbr: String = "",
    val name: String = "",
    val shortName: String = "",
    val logo: String = "",
) {
    /** The name a line uses: the short name ("Otters"), else the abbreviation, else the full name. */
    val displayName: String get() = shortName.ifBlank { abbr.ifBlank { name } }

    fun toTeam(): TallyTeam = TallyTeam(id = id, abbr = abbr, name = name, shortName = shortName, logo = logo)
}

@Serializable
data class DvrJobGame(
    val id: String = "",
    val league: String = "",
    val leaguePath: String = "",
    /** ISO-8601 */
    val start: String = "",
    val away: DvrTeam = DvrTeam(),
    val home: DvrTeam = DvrTeam(),
)

@Serializable
data class DvrJob(
    val id: String = "",
    val ruleId: String = "",
    val state: String = "",
    val reason: String? = null,
    /** canceled | disk | maxLength | postponed | restart, when a recording stopped before the game's end. */
    val stopReason: String? = null,
    /** "Away at Home" */
    val title: String = "",
    val game: DvrJobGame = DvrJobGame(),
    val createdAt: String? = null,
    val startedAt: String? = null,
    val endedAt: String? = null,
    val channelName: String? = null,
    /** Bytes recorded so far. */
    val bytes: Long = 0,
    /** Seconds recorded so far (the length of a finished recording). */
    val seconds: Double = 0.0,
    val fileBytes: Long = 0,
    val itemId: String? = null,
    /** Root-relative, signed start-over playlist, while recording. */
    val startOverPath: String? = null,
    /**
     * Whether the recording can be played from a library yet: `ready` ([itemId] is set), `adding` (a library covers
     * the file, Jellyfin has not picked it up yet) or `noLibrary` (no library covers the recordings folder). Older
     * plugins leave it out: see [libraryAvailability].
     */
    val libraryState: String? = null,
) {
    val isRecording: Boolean get() = state == DvrState.RECORDING
    val isPending: Boolean get() = state == DvrState.SCHEDULED || state == DvrState.WAITING
    val isFinal: Boolean get() = state == DvrState.DONE || state == DvrState.FAILED || state == DvrState.CANCELED
}

/** `GET /JellyTV/Client/v1/recordings/storage[?gameId=]`. */
@Serializable
data class DvrStorage(
    val freeBytes: Long? = null,
    val totalBytes: Long? = null,
    val usedBytes: Long = 0,
    val reserveBytes: Long = 0,
    val estimate: DvrEstimate? = null,
)

@Serializable
data class DvrEstimate(
    val gameId: String = "",
    val bytes: Long = 0,
    val fits: Boolean = true,
    /** Why it won't fit ("Not enough space: needs ~9 GB, 4 GB free"), from the server. */
    val message: String? = null,
)

/** Why a finished recording cannot be played yet, or [READY]: the job's `libraryState`. */
enum class LibraryAvailability {
    READY,
    ADDING,
    NO_LIBRARY,
}

/**
 * The job's [DvrJob.libraryState], with the rules for plugins that do not send it: a recording with an `itemId` is
 * ready; one without is still being added. An unknown value counts the same way.
 */
val DvrJob.libraryAvailability: LibraryAvailability
    get() =
        when {
            itemId != null -> LibraryAvailability.READY
            libraryState == "noLibrary" -> LibraryAvailability.NO_LIBRARY
            else -> LibraryAvailability.ADDING
        }

/** The job states, as the server names them. */
object DvrState {
    const val SCHEDULED = "scheduled"
    const val WAITING = "waiting"
    const val RECORDING = "recording"
    const val FINISHING = "finishing"
    const val DONE = "done"
    const val FAILED = "failed"
    const val CANCELED = "canceled"
}

/**
 * What the app shows for one game's recording: the list's job when the list has one (it knows when the recording
 * started), else the board's [TallyGame.recording].
 */
data class GameRecordingView(
    val state: String,
    val jobId: String,
    val startedAt: String? = null,
    val startOverPath: String? = null,
    val itemId: String? = null,
    val reason: String? = null,
    val seconds: Double = 0.0,
) {
    val isRecording: Boolean get() = state == DvrState.RECORDING
    val isPending: Boolean get() = state == DvrState.SCHEDULED || state == DvrState.WAITING
    val isFailed: Boolean get() = state == DvrState.FAILED
    val isFinal: Boolean get() = state == DvrState.DONE || state == DvrState.FAILED || state == DvrState.CANCELED

    /** A game can be recorded (again) when it has no job, or only one that ended without recording it. */
    val allowsNewRecording: Boolean get() = state == DvrState.FAILED || state == DvrState.CANCELED
}

/** The most relevant job of [gameId]: one still running first, then a finished recording, then the newest. */
fun List<DvrJob>.forGame(gameId: String): DvrJob? =
    filter { it.game.id == gameId }
        .sortedWith(
            compareBy<DvrJob> { if (it.isFinal) 1 else 0 }
                .thenBy { if (it.state == DvrState.DONE) 0 else 1 }
                .thenByDescending { it.createdAt ?: "" },
        ).firstOrNull()

/** [game]'s recording from the list when it has the game's job, else from the board. */
fun recordingView(
    game: TallyGame,
    list: DvrList?,
): GameRecordingView? {
    val job = list?.jobs?.forGame(game.id)
    if (job != null) {
        return GameRecordingView(
            state = job.state,
            jobId = job.id,
            startedAt = job.startedAt,
            startOverPath = job.startOverPath ?: game.recording?.startOverPath?.takeIf { job.isRecording },
            itemId = job.itemId,
            reason = job.reason,
            seconds = job.seconds,
        )
    }
    val board = game.recording ?: return null
    return GameRecordingView(
        state = board.state,
        jobId = board.jobId,
        startOverPath = board.startOverPath,
        itemId = board.itemId,
        reason = board.reason,
    )
}

/**
 * No spoilers: a finished game that has (or is making) a recording keeps its score and result out of sight. The
 * board's own field is enough, so every card can decide without the recordings list.
 */
val TallyGame.hasRecordingToWatch: Boolean
    get() =
        recording?.state.let {
            it == DvrState.DONE || it == DvrState.FINISHING || it == DvrState.RECORDING
        }

val TallyGame.spoilerGuarded: Boolean get() = isFinal && hasRecordingToWatch

/**
 * The team rule that records [team]'s games in [game]'s league, if any. Rules name the league by its path
 * ("baseball/mlb") and the board by its label ("MLB"): a job of the same league tells which path that is; without
 * one, the sport the path starts with has to do (team ids are only unique within a league).
 */
fun DvrList.teamRuleFor(
    game: TallyGame,
    team: TallyTeam,
): DvrRule? {
    if (team.id.isBlank()) return null
    val knownPath =
        jobs
            .firstOrNull { it.game.league.equals(game.league, ignoreCase = true) && it.game.leaguePath.isNotBlank() }
            ?.game
            ?.leaguePath
    return rules.firstOrNull { rule ->
        rule.isTeam &&
            rule.teamId == team.id &&
            when {
                rule.leaguePath.isBlank() -> true
                knownPath != null -> rule.leaguePath.equals(knownPath, ignoreCase = true)
                else -> rule.leaguePath.substringBefore('/').equals(game.sport, ignoreCase = true)
            }
    }
}
